src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp

```diff
@@ -0,0 +1,258 @@
+            const TypeSupport& type)
+        : limits_(pool_config.initial_size,
+                pool_config.maximum_size ? pool_config.maximum_size : std::numeric_limits<size_t>::max(),
+                1)
+        , free_loans_(limits_)
+        , used_loans_(limits_)
+        , type_(type)
+    {
+        for (size_t n = 0; n < limits_.initial; ++n)
+        {
+            OutstandingLoanItem item;
+            if (!type_->is_plain())
+            {
+                item.sample = type_->createData();
+            }
+            free_loans_.push_back(item);
+        }
+    }
+
+    ~SampleLoanManager()
+    {
+        if (!type_->is_plain())
+        {
+            for (const OutstandingLoanItem& item : free_loans_)
+            {
+                type_->deleteData(item.sample);
+            }
+        }
+    }
+
+    int32_t num_allocated() const
+    {
+        assert(used_loans_.size() <= std::numeric_limits<int32_t>::max());
+        return static_cast<int32_t>(used_loans_.size());
+    }
+
+    void get_loan(
+            CacheChange_t* change,
+            void*& sample)
+    {
+        // Early return an already loaned item
+        OutstandingLoanItem* item = find_by_change(change);
+        if (nullptr != item)
+        {
+            item->num_refs += 1;
+            sample = item->sample;
+            return;
+        }
+
+        // Get an item from the pool
+        if (free_loans_.empty())
+        {
+            // Try to create a new entry
+            item = used_loans_.push_back({});
+            if (nullptr != item)
+            {
+                // Create sample if necessary
+                if (!type_->is_plain())
+                {
+                    item->sample = type_->createData();
+                }
+            }
+        }
+        else
+        {
+            // Reuse a free entry
+            item = used_loans_.push_back(free_loans_.back());
+            assert(nullptr != item);
+            free_loans_.pop_back();
+        }
+
+        // Should always find an entry, as resource limits are checked before calling this method
+        assert(nullptr != item);
+
+        // Should be the first time we loan this item
+        assert(item->num_refs == 0);
+
+        // Increment references of input payload
+        CacheChange_t tmp;
+        tmp.copy_not_memcpy(change);
+        item->owner = change->payload_owner();
+        change->payload_owner()->get_payload(change->serializedPayload, item->owner, tmp);
+        item->owner = tmp.payload_owner();
+        item->payload = tmp.serializedPayload;
+        tmp.payload_owner(nullptr);
+        tmp.serializedPayload.data = nullptr;
+
+        // Perform deserialization
+        if (type_->is_plain())
+        {
+            auto ptr = item->payload.data;
+            ptr += item->payload.representation_header_size;
+            item->sample = ptr;
+        }
+        else
+        {
+            type_->deserialize(&item->payload, item->sample);
```

Comment:
Not tested

Reply:
I will add tests for non-plain types on #1673